### ISSUE TITLE
timezone information is dropped and everything is displayed in UTC for timezoned data

### ISSUE THREAD

**Issue by fboerman** [evidence-dev/evidence#1652]
## Steps To Reproduce

1. create a query which returns data with at least one column with timestamp with timezone data, make sure the data is in a non UTC timezone
3. load it in a page
4. put it in a datatable and see that data is in UTC format

in the same category if you use a timestamp in a where clause with timezone info no data is returned.

### Environment

- Node version (`node -v`): v20.11.1
- npm version (`npm -v`): 10.4.0
- OS: archlinux 6.6.10-arch1-1-surface  SMP PREEMPT_DYNAMIC Sat, 13 Jan 2024 00:20:37 +0000 x86_64 GNU/Linux
- Browser: firefox
- Database: postgresql

## Expected Behavior

data is in the timezone that the source was in

## Actual Behaviour

data is shown in UTC

## Workarounds

to get both the visualization as well as the filtering working it is needed to load the icu extension and then do a double timezone conversion. steps to do this:
1. add ```"@duckdb/duckdb-wasm": "1.28.1-dev148.0"``` to overrides in packages.json and install
2. add at the top of your page the following snippit: ```<script>
import {query} from "@evidence-dev/universal-sql/client-duckdb";
query(" LOAD icu; ")
</script>```
3.  adjust your filter to use for example ```time >= TIMESTAMPTZ '2024-01-01 00:00:00 Europe/Amsterdam'```
4. double convert your time column like this: ```"time" at time zone 'UTC' at time zone 'Europe/Amsterdam' as time```

now the data should be correctly filtered and shown
WARNING: this newer version of duckdb-wasm has not been fully tested with evidence! it seems to work but behold for edge cases.

### Drawbacks to workaround:
when clicking download data from for example datatable or linechart the data in the csv is shown as being UTC (with the Z) time but that is wrong, its actually in the local timezone specified. this is actually *wrong* instead of different dispaly then you want so be aware

## possible ideas for feature to fix this
as a feedback point it would be great if the timezone shown on a page would be configurable, both as a global default as well as a possible override per page or query

**Comment by jaroet** [comment]
I just wanted to add that I too am running into this issue. My case is specifically with a dateRange component to filter out data. That goes wrong. This makes it for me impossible to use Evidence as almost all our reports have or will have a date based filter. 

I do not have any specific timestamptz type of fields in my table but regular timestamp. But as the timezone of the postgresql server I am using is Europe/Amsterdam I guess all timestamp fields are in fact timestamptz.

**Comment by hilkeheremans** [comment]
Just to add some info here, here's some more info about the workaround above:
- you need to add an override in your package.json to use a later version of DuckDB WASM than Evidence currently installs
- 'add at the top of your page' --> I added it in a custom `+layout.svelte` file (as in https://docs.evidence.dev/reference/themes-and-layouts/ )
- the double conversion can be made a little more livable by also adding a macro to the layout as follows and using that
```
    query(
        "CREATE FUNCTION tz_fix(x, b := 'Europe/Amsterdam') AS (x::TIMESTAMP at time zone 'UTC' at time zone b);",
    );
``` 
and then using that for double conversions in time columns like so `tz_fix("time")`. Also lets you optionally use other time zones in case of need

The handy part about using a macro here is that it should be easier to 'remove' the fix in various places as soon as Evidence has a more permanent solution, simply by changing the function to a no-op

**Comment by hilkeheremans** [comment]
Just to add again, I found that Evidence refuses to make a production build (weird 500 error) if you attempt this hack. The root cause of this is build behavior vs dev behavior, where the timing of the query calls seems to be different.

The way to fix that is to force rendering to wait until the extensions are actually loaded, by adjusting your `+layout.svelte` something like this. **Note that the workaround below can be applied to basically anything async that you want to wait for before any page load.**

```
<script>
    import "@evidence-dev/tailwind/fonts.css";
    import "../app.css";
    import { EvidenceDefaultLayout } from "@evidence-dev/core-components";
    import { query } from "@evidence-dev/universal-sql/client-duckdb";
    async function installICU() {
        await query(" LOAD icu; ");
        await query(
            "CREATE FUNCTION tz_fix(x, b := 'Europe/Amsterdam') AS (x::TIMESTAMP at time zone 'UTC' at time zone b);",
        );
    }
    export let data;
</script>

{#await installICU()}
    <p>Loading</p>
{:then result}
    <EvidenceDefaultLayout {data}>
        <slot slot="content" />
    </EvidenceDefaultLayout>
{:catch error}
    <p style="color: red">{error.message}</p>
{/await}
```

**Comment by hilkeheremans** [comment]
Another addition: use 

```
CREATE OR REPLACE FUNCTION
```

instead of just

```
CREATE FUNCTION
```

to maintain stability with HMR

**Comment by hilkeheremans** [comment]
Another addition:

Unfortunately, this fix does not work well. It breaks (at least) the following functionality on build (not in dev mode):
- templates pages do not render
- all-queries.json + arrow files do not pregenerate (=query result caching), causing a performance drop in the built dashboard

Use at your own risk!

**Comment by archiewood** [comment]
I have taken some time to chase this down.

Evidence has several stages of it's data pipeline, so I wanted to ensure we know what we are doing at each stage.

## Stages

`npm run sources`
1. Data in the native DB (eg Postgres, DuckDB), using native DB types
2. Data gets read by the native nodejs connector, and converted into JS types
3. Data is mapped from JS types into a set of 4 evidence types (string, number, bool, datetime)
4. Data is converted from an JS Array into an arrow vector, columnwise, and then into an arrow table
5. Data is written from arrow table into parquet file on disk

`npm run dev`
6. DuckDB wasm reads data from parquet file based on markdown query
7. Data is saved in Query Store
8. Components read data from query store (back to JS array)
9. Components format data and display it

## Complications
- JS does not have a concept of timezone aware timestamps. All timestamps are stored as milliseconds since 1970, in UTC.
- Some "databases" do not have a native concept of time (eg SQLite, CSV) and use local time strings to represent time
- Some developers choose (or use systems that choose) to store data in timestamps without timezones, but with an _implicit_ timezone choice of their local TZ. If you change the system timezone it can changes the implied timestamp

## Evidence Today

I have run an end to end test using DuckDB to confirm current behaviour.
1. Evidence takes values as they come out of queries, and doesn't change them.
2. However, some(!) components choose to render these values in UTC. Specifically the DataTable, and the QueryViewer
3. If they don't have a timezone specified, it assumes they are UTC

## Fixes

- For databases/sources that do not have timezone support, we should give people the option to configure a timezone.
- The DataTable and QueryViewer should not silently convert dates to UTC
- (Eventually) We should support allowing users to change the display timezone to one other than their browsers'

**Comment by fboerman** [comment]
Hi Archie,

Thanks for the write up. Perhaps good to mention that in order to get timezone support in duckdb you need to load ICU extention. If you do that you are already 90% there

-------- Original Message --------

**Comment by archiewood** [comment]
This should now be unblocked with the latest release of DuckDB WASM

**Comment by benyn** [comment]
I see #2163 was closed. Is there another plan to address time zone issues?

Loading the DuckDB `icu` extension in `+layout.svelte` mostly works but it'd be better if we don't need to show the "Loading" text.

**Comment by EthanGoldman** [comment]
I'd like to add a vote for addressing this issue, hopefully in a way that supports tz-aware timestamps being preserved from the original parquet or duckdb source files. I'm working with data that spans multiple time zones, so I can't just convert the UTC data to my local time zone (which is not even one of the correct time zones!)

This bug is currently resulting in charts that show solar panel energy generation peaking in the middle of the night, which is bound to degrade confidence in our application. 

As a workaround, can I store the timestamp data as a string with the UTC offset included and then convert it back to a tz-aware timestamp as part of the query, or am I better off creating a tz-naive timestamp and just dealing with the mess around DST transitions?

**Comment by fboerman** [comment]
hi @EthanGoldman  I also do not understand why something so fundamental is still not being adressed tbh. I am currently doing the trick to load in string and then convert back which works well as long as you manually take care of sorting issues. @archiewood any update on this?

**Comment by archiewood** [comment]
hi all, understand your frustration on this - we are aware of the issue, but have not yet addressed it

to fix properly is quite a fundamental, far reaching and potentially breaking change, so we want to make sure we have enough bandwidth before we take it on

**Comment by fboerman** [comment]
hi @archiewood  thanks for the quick reply. is there any timeline on this possible? Thanks!

**Comment by archiewood** [comment]
No timeline as yet

**Comment by pniederlag** [comment]
Preface: Evidence is such a great product and really fun to use

When it comes to internatonalization it unfortunatly yields small and big problems. While number(currency) formatting is tolerable handling of datetimes in an international context and/or with daylight-savings (timezone related information) seems to be a crucial feature for true professional usage.

I built my use-case with a duckdb "testcase" to ensure my data is handled correct. However as i now use this very well tested duckdb data inside evidence I have to learn that my queries in evidence yield broken/unexpected results. :-<

My native (duckdb) checkup SQL

```
SELECT
    timezone('Europe/Berlin', tz_datetime) AS date_in_europe_berlin,
    timezone('UTC', tz_datetime) AS date_in_utc,
    strftime(tz_datetime , '%H') as hour_by_strftime,
    date_part('hour', tz_datetime) as hour_by_date_part,
    date_part('timezone_hour', tz_datetime) as hour_offset_by_date_part,
    tz_datetime,
    tz_offset as offset_as_hardcoded_string_for_checkup
FROM
   dummy.main.dummy
```

My native (duckdb) results:

|date_in_europe_berlin|date_in_utc|hour_by_strftime|hour_by_date_part|hour_offset_by_date_part|tz_datetime|offset_as_hardcoded_string_for_checkup|
|---------------------|-----------|----------------|-----------------|------------------------|-----------|--------------------------------------|
|2025-01-02 18:00:00.000|2025-01-02 17:00:00.000|18|18|1|2025-01-02 17:00:00.000 +0000|+0100|
|2025-04-02 18:00:00.000|2025-04-02 16:00:00.000|18|18|2|2025-04-02 16:00:00.000 +0000|+0200|

these are quite correct and reasonable for me :->

However, when shifting this to evidence I get broken/unwanted results

- hour_by_strftime yields 17 / 16 (off by timezone diff)
- hour_by_date_part yields 17 / 16  (off by timezone diff)
- hour_offset_by_date_part yields 0 (merely wrong)
- function timezone is not availabel (not important, just a notable difference)

I'll probably do some workarounds by fiddling with the data and do the relevant math before pushing the data into evidence. However it would be very welcomed if  i18n support could be improved on these matters.

Maybe pushing local_time into a non TZ-field does the trick? I'll check that as well...

Thx for listening and such a nice product!

**Comment by hughess** [comment]
This issue should be solved in the new version of Evidence Core: https://docs.evidence.dev/
